In [207]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import string
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, StandardScaler
from moscow_housing.display_data import import_data

plt.style.use('ggplot')
%matplotlib inline

In this notebook we will do EDA (explanatory data analysis). We will look at dataset standard statistics, individual feature analysis and finding any relations or trends considering multiple features. As a conlusion we will suggest features which can be added as a result of EDA

In [208]:
#import data
data_train, data_test = import_data()

#sometimes we have to look at all the data togheter
all_data = pd.concat([data_train,data_test])
all_data

id  seller       price  area_total  area_kitchen  area_living  floor  \
0         0     3.0   7139520.0        59.2          12.5         31.0    2.0   
1         1     NaN  10500000.0        88.0          14.2         48.0   18.0   
2         2     3.0   9019650.0        78.5          22.5         40.8   12.0   
3         3     NaN  10500000.0        88.0          14.0         48.0   18.0   
4         4     NaN  13900000.0        78.0          17.0         35.0    7.0   
...     ...     ...         ...         ...           ...          ...    ...   
9932  33217     3.0         NaN       106.0          19.9         56.7   16.0   
9933  33218     NaN         NaN        82.0           NaN          NaN    3.0   
9934  33219     NaN         NaN        49.3           NaN          NaN   15.0   
9935  33220     NaN         NaN        38.8          10.5         15.1   14.0   
9936  33221     1.0         NaN        71.5           7.0         49.0    2.0   

      rooms  layout  ceiling  ...  address  constructed  material  stories  \
0       2.0     NaN     2.65  ...   к2.5/2       2021.0       3.0      9.0   
1       3.0     1.0      NaN  ...     14к3       2010.0       3.0     25.0   
2       3.0     NaN     2.65  ...       38       2021.0       3.0     15.0   
3       3.0     NaN      NaN  ...     14к3       2010.0       3.0     25.0   
4       2.0     1.0     2.90  ...      1к3       2017.0       2.0     15.0   
...     ...     ...      ...  ...      ...          ...       ...      ...   
9932    3.0     NaN     3.30  ...      10А       2020.0       2.0     20.0   
9933    3.0     NaN      NaN  ...       к1       2021.0       2.0     20.0   
9934    1.0     NaN     0.00  ...     70к5       2016.0       2.0     24.0   
9935    1.0     NaN     3.30  ...        1       2019.0       NaN     14.0   
9936    3.0     NaN      NaN  ...        2          NaN       NaN      2.0   

      elevator_without  elevator_passenger  elevator_service  parking  \
0                  0.0                 1.0               1.0      1.0   
1                  0.0                 1.0               1.0      1.0   
2                  0.0                 1.0               1.0      1.0   
3                  0.0                 1.0               1.0      1.0   
4                  0.0                 1.0               1.0      1.0   
...                ...                 ...               ...      ...   
9932               0.0                 1.0               1.0      0.0   
9933               0.0                 1.0               1.0      1.0   
9934               0.0                 1.0               1.0      0.0   
9935               1.0                 1.0               1.0      0.0   
9936               NaN                 NaN               NaN      NaN   

      garbage_chute  heating  
0               NaN      NaN  
1               NaN      0.0  
2               NaN      NaN  
3               NaN      0.0  
4               0.0      0.0  
...             ...      ...  
9932            NaN      0.0  
9933            NaN      NaN  
9934            1.0      1.0  
9935            1.0      NaN  
9936            NaN      NaN  

[33222 rows x 34 columns]

In [209]:
#Now lets look at all the basics of the data
print('Number of Training Examples = {}'.format(data_train.shape[0]))
print('Number of Test Examples = {}\n'.format(data_test.shape[0]))
print('Training X Shape = {}'.format(data_train.shape))
print('Training y Shape = {}\n'.format(data_train['price'].shape[0]))
print('Test X Shape = {}'.format(data_test.shape))
print('Test y Shape = {}\n'.format(data_test.shape[0]))
print('Train columns \n', list(data_train.columns))
print('Test columns \n', list(data_test.columns))

#check type of each column, and see how many duplicates we have:
for column in data_train.columns:
    print('\n - type of column: ', data_train.dtypes[column],'\n - number of null values: ', data_train[column].isnull().sum(), '\n - number of unique inputs: ', data_train[column].value_counts().count(),'\n', data_train[column].value_counts())


Number of Training Examples = 23285
Number of Test Examples = 9937

Training X Shape = (23285, 34)
Training y Shape = 23285

Test X Shape = (9937, 33)
Test y Shape = 9937

Train columns 
 ['id', 'seller', 'price', 'area_total', 'area_kitchen', 'area_living', 'floor', 'rooms', 'layout', 'ceiling', 'bathrooms_shared', 'bathrooms_private', 'windows_court', 'windows_street', 'balconies', 'loggias', 'condition', 'phones', 'building_id', 'new', 'latitude', 'longitude', 'district', 'street', 'address', 'constructed', 'material', 'stories', 'elevator_without', 'elevator_passenger', 'elevator_service', 'parking', 'garbage_chute', 'heating']
Test columns 
 ['id', 'seller', 'area_total', 'area_kitchen', 'area_living', 'floor', 'rooms', 'layout', 'ceiling', 'bathrooms_shared', 'bathrooms_private', 'windows_court', 'windows_street', 'balconies', 'loggias', 'condition', 'phones', 'building_id', 'new', 'latitude', 'longitude', 'district', 'street', 'address', 'constructed', 'material', 'stories', 'el

As we can see, the training and test set have the same number of columns except for price. Lets look at little bit closer at these features.
### **Features description**
* `id` - Unique ID for apartment
    - type of column:  int64
    - number of null values:  0
    - number of unique inputs:  23285
* `seller` - The type of apartment seller
 - type of column:  float64
 - number of null values:  8830
 - number of unique inputs:  4
* `price` - The listed price of the apartment (TARGET: only available in train)
 - type of column:  float64
 - number of null values:  0
 - number of unique inputs:  9504
* `area_total` - Total area of the apartment
 - type of column:  float64
 - number of null values:  0
 - number of unique inputs:  3249
* `area_kitchen` - Total kitchen area in the apartment
 - type of column:  float64
 - number of null values:  4721
 - number of unique inputs:  385
* `area_living` - Total living space area in the apartment
 - type of column:  float64
 - number of null values:  3882
 - number of unique inputs:  1041
* `floor`- Primary building floor of the apartment
 - type of column:  float64
 - number of null values:  0
 - number of unique inputs:  84
* `rooms` - Number of rooms in the apartment
 - type of column:  float64
 - number of null values:  0
 - number of unique inputs:  6
* `layout` - Overal apartment layout
 - type of column:  float64
 - number of null values:  17642
 - number of unique inputs:  3
* `ceiling` - Ceiling height in the apartment
 - type of column:  float64
 - number of null values:  11093
 - number of unique inputs:  121
* `bathrooms_shared` - Number of shared bathrooms
 - type of column:  float64
 - number of null values:  3872
 - number of unique inputs:  5
* `bathrooms_private` - Number of private bathrooms
 - type of column:  float64
 - number of null values:  3872
 - number of unique inputs:  5
* `windows_court` - Whether the apartment has windows facing a courtyard
 - type of column:  float64
 - number of null values:  8072
 - number of unique inputs:  2
* `windows_street` - Whether the apartment has windows facing a street
 - type of column:  float64
 - number of null values:  8072
 - number of unique inputs:  2
* `balconies` - Number of balconies in the apartment
 - type of column:  float64
 - number of null values:  10520
 - number of unique inputs:  5
* `loggias`- Number of loggias in the apartment (balcony-like, google it)
 - type of column:  float64
 - number of null values:  10520
 - number of unique inputs:  5
* `condition` - Current condition of the apartment
 - type of column:  float64
 - number of null values:  9810
 - number of unique inputs:  4
* `phones` - Number of phone numbers associated with the advert
 - type of column:  float64
 - number of null values:  80
 - number of unique inputs:  3
* `building_id` - ID used to map apartments to buildings
 - type of column:  int64
 - number of null values:  0
 - number of unique inputs:  6791
* `new` - Whether it is an old or new building
 - type of column:  float64
 - number of null values:  264
 - number of unique inputs:  2
* `latitude` - Latitude coordinate of building
 - type of column:  float64
 - number of null values:  0
 - number of unique inputs:  6208
* `longitude` - Longitude coordinate of building
 - type of column:  float64
 - number of null values:  0
 - number of unique inputs:  6159
* `district` - Administriative district within Moscow
 - type of column:  float64
 - number of null values:  130
 - number of unique inputs:  12
* `street` - Bulding street name
 - type of column:  object
 - number of null values:  0
 - number of unique inputs:  1682
* `address` - Building address (within street)
 - type of column:  object
 - number of null values:  0
 - number of unique inputs:  1851
* `constructed` - Year when the building was constructed
 - type of column:  float64
 - number of null values:  794
 - number of unique inputs:  128
* `material` - Primary building material used in building
 - type of column:  float64
 - number of null values:  3972
 - number of unique inputs:  7
* `stories` - Total number of floors in the building
 - type of column:  float64
 - number of null values:  0
 - number of unique inputs:  62
* `elevator_without` - Whether the building has apartments without elevator access
 - type of column:  float64
 - number of null values:  357
 - number of unique inputs:  2
* `elevator_passenger` - Whether the building has apartments with passenger elevator access
 - type of column:  float64
 - number of null values:  357
 - number of unique inputs:  2
* `elevator_service` - Whether the building has apartments with service elevator access
 - type of column:  float64
 - number of null values:  357
 - number of unique inputs:  2
* `parking` - Parking options for building
 - type of column:  float64
 - number of null values:  6788
 - number of unique inputs:  3
* `garbage_chute` - Whether the building has a garbage chute system
 - type of column:  float64
 - number of null values:  8811
 - number of unique inputs:  2
* `heating` - Primary heating system used in the building
 - type of column:  float64
 - number of null values:  8161
 - number of unique inputs:  4


Building - building average price som feature
Latitude-longtiude - avstand fra sentrum
distict - district average price som feature
street og address -

kategorisek fylle inn nan?

In [210]:
#Nå som dataen er lastet inn og vi har sett på antall verdier osv er det på tide å gå nærmere inn på hver
#feature

#sjekker korrelelasjon for alle features for å se om noen er spennede
corr = data_train.corr()
corr.style.background_gradient(cmap='coolwarm')

### **Grouping**
We can see that longitude and latitude has very low correleation with price. However when plotted on a map it seems to be a clear correleation
Lets try and change these into radius instead.

In [211]:
data_with_radius = data_train
data_with_radius['radius'] = np.sqrt((data_with_radius['latitude']-55.75)**2 + (data_with_radius['longitude']-37.55)**2)

data_with_groups = data_with_radius

corr_2 = data_with_radius.corr()
corr_2.style.background_gradient(cmap='coolwarm')

It is clear that radius correlates much more with price than longitude and latitude, therefore we created a better feature than previous.
Now i want to look at if it is possible to group some of the data into groups, and see if this gives additional information

In [212]:
#create groups and find mean for each group
grouped_by_districts = data_train.groupby(['district'])
mean_price = grouped_by_districts['price'].mean()

data_with_group = pd.merge(data_train,mean_price, on='district', how='left')

data_with_groups['district_price'] = data_with_group['price_y']

#check correleation now
corr_3 = data_with_groups.corr()
corr_3.style.background_gradient(cmap='coolwarm')

This new feature does also correlate more with the price than the previous feature
Lets see if we can do something similar with street

In [213]:
#create groups and find mean for each group
grouped_by_districts = data_train.groupby(['street'])
mean_price = grouped_by_districts['price'].mean()
print(mean_price)

data_with_group = pd.merge(data_train,mean_price, on='street', how='left')

data_with_groups['street_price'] = data_with_group['price_y']

#check correleation now
corr_4 = data_with_groups.corr()
corr_4.style.background_gradient(cmap='coolwarm')

street
1-й Амбулаторный проезд       7.650000e+06
1-й Басманный переулок        4.040874e+07
1-й Боткинский проезд         4.113333e+07
1-й Волоколамский проезд      8.910000e+06
1-й Грайвороновский проезд    9.207602e+06
                                  ...     
улица Ясеневая                1.741286e+07
улица Ясная                   8.283333e+06
шоссе Обводное                7.500000e+06
шоссе Энтузиастов             1.059866e+07
№ 211 кв-л                    4.300000e+06
Name: price, Length: 1682, dtype: float64


this one also correlated very much, starting to question if this is correct approach

In [214]:
#create groups and find mean for each group
grouped_by_districts = data_train.groupby(['building_id'])
mean_price = grouped_by_districts['price'].mean()
print(mean_price)

data_with_group = pd.merge(data_train,mean_price, on='building_id', how='left')

data_with_groups['building_price'] = data_with_group['price_y']

#check correleation now
corr_5 = data_with_groups.corr()
corr_5.style.background_gradient(cmap='coolwarm')

building_id
0       1.130000e+07
1       7.100000e+06
2       1.360000e+07
3       1.010000e+07
7       5.794500e+07
            ...     
9730    9.150000e+06
9731    3.190000e+07
9732    1.015000e+07
9733    8.000000e+07
9735    8.277614e+06
Name: price, Length: 6791, dtype: float64


Now we can see that street price and building price correlates very much, which means that one of them have to go.
We think building_price is better because there is more equal within a building than within a street.

## **Outliers**

Now lets look at the outliers, lets start looking at the radius

In [215]:
max_value = data_with_groups['radius'].max()
min_value = data_with_groups['radius'].min()

data_with_radius = data_test
data_with_radius['radius'] = np.sqrt((data_with_radius['latitude']-55.75)**2 + (data_with_radius['longitude']-37.55)**2)

print('rows with missing latitude and longitude in test data: \n ',data_with_radius[['longitude','latitude']][data_with_radius['longitude'].isnull() == True])

max_value_test = data_with_radius['radius'].nlargest(10)
print('max radius from trainingset: \n',max_value, '\n10 highest values from test set:\n',max_value_test)

rows with missing latitude and longitude in test data: 
      longitude  latitude
23        NaN       NaN
90        NaN       NaN
max radius from trainingset: 
 0.7564444961376798 
10 highest values from test set:
 2511    106.579234
5090    106.579234
6959    106.579234
8596    106.579234
4719     96.021654
9547     96.012910
2529     39.132054
3117      0.756266
834       0.755604
873       0.755604
Name: radius, dtype: float64


As we can see from the code above, the test set is missing 2 longitude/latitude entries
there is also 6 houses which is outsie moscow, have to remove radius for these buildings and look at other factors
Want to look at the rows which is missing longitude and latitude, do they have anything else which connects the to moscow?

row 23 and 90 is missing radius

rows outside moscov:
2511    106.579234
5090    106.579234
6959    106.579234
8596    106.579234
4719     96.021654
9547     96.012910
2529     39.132054

In [216]:
data_test = data_with_radius

#print(data_with_radius.loc[[23]]['street'])
#print(data_with_radius.loc[[90]]['building_id'])
#print(data_with_radius[data_with_radius['street']=='пос. Коммунарка'])
data_test._set_value(23,'radius',0.203899)
data_test._set_value(90,'radius',0.203899)



#Now we dont have any missing radius values in the test data
#lets look at the rows outside of moscov
data_test._set_value(2511,'radius',0.218159)
data_test._set_value(5090,'radius',0.218159)
data_test._set_value(6959,'radius',0.218159)
data_test._set_value(8596,'radius',0.218159)

#print(data_test.loc[[2511]][['street']])
#print(data_test.loc[[5090]][['longitude','latitude']])
#print(data_test.loc[[6959]])
#print(data_test.loc[[8596]])
#print(data_with_radius[data_with_radius['street']=='Бунинские Луга ЖК'])

#print(data_test.loc[[4719]][['street','address']])
#print(data_with_radius[['address','radius']][data_with_radius['street']=='улица Центральная'])

data_test._set_value(4719, 'radius',0.19580)
data_test._set_value(9547, 'radius',0.19520)

#print(data_test.loc[[2529]][['street','address','district']])
#print(data_with_radius[['longitude','latitude']][data_with_radius['street']=='улица 1-я Линия'])
data_test._set_value(2529, 'radius', np.sqrt((37.464994-37.55)**2+(55.627666-55.75)**2))